servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java

```diff
@@ -0,0 +1,547 @@
+        }
+        this.maxMappedDemand = maxMappedDemand;
+        this.maxConcurrency = maxConcurrency;
+        this.delayError = delayError;
+    }
+
+    @Override
+    public Subscriber<? super T> apply(final Subscriber<? super R> subscriber) {
+        return new FlatMapSubscriber<>(this, subscriber);
+    }
+
+    private static final class FlatMapSubscriber<T, R> implements Subscriber<T>, Subscription {
+        private static final Object MAPPED_SOURCE_COMPLETE = new Object();
+        @SuppressWarnings("rawtypes")
+        private static final AtomicReferenceFieldUpdater<FlatMapSubscriber, CompositeException> delayedErrorUpdater =
+                newUpdater(FlatMapSubscriber.class, CompositeException.class, "delayedError");
+        @SuppressWarnings("rawtypes")
+        private static final AtomicIntegerFieldUpdater<FlatMapSubscriber> emittingLockUpdater =
+                AtomicIntegerFieldUpdater.newUpdater(FlatMapSubscriber.class, "emittingLock");
+        @SuppressWarnings("rawtypes")
+        private static final AtomicLongFieldUpdater<FlatMapSubscriber> pendingDemandUpdater =
+                AtomicLongFieldUpdater.newUpdater(FlatMapSubscriber.class, "pendingDemand");
+        @SuppressWarnings("rawtypes")
+        private static final AtomicIntegerFieldUpdater<FlatMapSubscriber> activeMappedSourcesUpdater =
+                AtomicIntegerFieldUpdater.newUpdater(FlatMapSubscriber.class, "activeMappedSources");
+        @SuppressWarnings("rawtypes")
+        private static final AtomicReferenceFieldUpdater<FlatMapSubscriber, TerminalNotification>
+                terminalNotificationUpdater = newUpdater(FlatMapSubscriber.class, TerminalNotification.class,
+                "terminalNotification");
+
+        @SuppressWarnings("unused")
+        @Nullable
+        private volatile TerminalNotification terminalNotification;
+        @Nullable
+        private volatile CompositeException delayedError;
+        @SuppressWarnings("unused")
+        private volatile int emittingLock;
+        private volatile int activeMappedSources;
+        private volatile long pendingDemand;
+
+        // protected by emitting lock, or only accessed inside the Subscriber thread
+        private boolean targetTerminated;
+        private int upstreamDemand;
+        @Nullable
+        private Subscription subscription;
+
+        private final Subscriber<? super R> target;
+        private final Queue<Object> signals;
+        private final PublisherFlatMapMerge<T, R> source;
+        private final Set<FlatMapPublisherSubscriber<T, R>> subscribers;
+
+        FlatMapSubscriber(PublisherFlatMapMerge<T, R> source, Subscriber<? super R> target) {
+            this.source = source;
+            this.target = target;
+            // Start with a small capacity as maxConcurrency can be large.
+            signals = newUnboundedMpscQueue(min(2, source.maxConcurrency * source.maxMappedDemand));
```

Comment:
This approach of child `Publisher`s requesting `maxMappedDemand` irrespective of downstream demand will tend to have larger buffers irrespective of how items are requested. 

An alternative approach could be to take a `minMappedDemand` and then split the downstream demand across `maxConcurrency` number of sources. 

- For `maxConcurrency == 8` and downstream demand `16`; we request `2` from each child `Publisher`.
- Each time a child `Publisher` emits as much as requested we split the remaining demand again in `maxConcurrency`. If `maxConcurrency > downstream demand` then request `minMappedDemand`.

This approach should result in smaller queues since we request depending on actual downstream demand as opposed to static `maxMappedDemand`. Additionally, we can default to lower defaults; `maxConcurrency: 8` and `minMappedDemand: 1`. The negative is that if we do not have as many items as `maxConcurrency` we end up requesting in smaller chunks.

WDYT?

Reply:
discussed offline and adopted this approach with trade-offs documented in javadocs